docs/source/core/core.rst

```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
+        assert input.compare(out2) == NeuralTypeComparisonResult.GREATER
+
+Custom element types
+~~~~~~~~~~~~~~~~~~~~
+It is possible to create user-defined element types to express the semantics of elements in your tensors.
+To do so, you need to inherit and implement abstract methods of `nemo.core.neural_types.elements.ElementType` class
+
+.. autoclass:: nemo.core.neural_types.elements.ElementType
+
+Note that element types can be parametrized. Consider this example where it distinguishes between audio sampled at 8Khz and 16Khz.
+
+.. code-block:: python
+
+    audio16K = NeuralType(axes=('B', 'T'), elements_type=AudioSignal(16000))
+    audio8K = NeuralType(axes=('B', 'T'), elements_type=AudioSignal(8000))
+
+    assert audio8K.compare(audio16K) == NeuralTypeComparisonResult.SAME_TYPE_INCOMPATIBLE_PARAMS
+    assert audio16K.compare(audio8K) == NeuralTypeComparisonResult.SAME_TYPE_INCOMPATIBLE_PARAMS
+
+Enforcing dimensions
+~~~~~~~~~~~~~~~~~~~~
+In addition to specifying tensor layout and elements' semantics, neural types also allow you to enforce tensor dimensions.
+You will have to use long notation to specify dimensions. Short notation only allows you to specify axes semantics and assumes
+arbitrary dimensions.
+
+.. code-block:: python
+
+        type1 = NeuralType(
+        (AxisType(AxisKind.Batch, 64), AxisType(AxisKind.Time, 10), AxisType(AxisKind.Dimension, 128)),
+        SpectrogramType(),
+        )
+        type2 = NeuralType(('B', 'T', 'C'), SpectrogramType())
+
+        # type2 will accept elements of type1 because their axes semantics match and type2 does not care about dimensions
+        assert type2.compare(type1), NeuralTypeComparisonResult.SAME
+        # type1 will not accept elements of type2 because it need dimensions to match strictly.
+        assert type1.compare(type2), NeuralTypeComparisonResult.DIM_INCOMPATIBLE
+
+Generic Axis kind
+~~~~~~~~~~~~~~~~~
+Sometimes (especially in the case of loss modules) it is useful to be able to specify a "generic" axis kind which will make it
+compatible with any other kind of axis. This is easy to express with Neural Types by using `nemo.core.neural_types.axes.AxisKind.Any` for axes.
+
+.. code-block:: python
+
+        type1 = NeuralType(('B', 'Any', 'Any'), SpectrogramType())
+        type2 = NeuralType(('B', 'T', 'C'), SpectrogramType())
+        type3 = NeuralType(('B', 'C', 'T'), SpectrogramType())
+
+        # type1 will accept elements of type2 and type3 because it only cares about element kind (SpectrogramType)
+        # number of axes (3) and that first one corresponds to batch
+        assert type1.compare(type2) == NeuralTypeComparisonResult.SAME
+        assert type1.compare(type3) == NeuralTypeComparisonResult.INCOMPATIBLE
+
+
+
+Container types
+~~~~~~~~~~~~~~~
+NeMo type system understands Python containers (lists). So if your module returns a list of typed tensors, the way to express it by
```

Comment:
returns a "nested" list - we do support it. Dunno if any model actually uses more than depth 1 nest though so this sentence is fine too.

Reply:
added